### ISSUE TITLE
Broken collision with invisible `CSGCombiners`

### ISSUE THREAD

**Issue by hoontee** [godotengine/godot#43251]
**Godot version:**
3.2.4.beta1 - master

**OS/device including version:**
Windows 10 2004

**Issue description:**
Invisible `CSGCombiners` with `use_collision = true` will not be collidable.
![image](https://user-images.githubusercontent.com/5272529/97960938-fa256300-1d77-11eb-876f-68d117e1714a.png)

**Minimal reproduction project:**
[Invisible CSGShape Collision.zip](https://github.com/godotengine/godot/files/5470133/Invisible.CSGShape.Collision.zip)

**Comment by Janders1800** [comment]
Yep, happens to me as well, they also trow null on _on_body_entered / exited.

**Comment by hoontee** [comment]
@madmiraal This issue must have been caused by #40919, but it's definitely fixed in #40814.

**Comment by madmiraal** [comment]
I think this is the desired behaviour i.e. it was fixed with #40919. If any of the components of a CSG shape are made invisible, they are no longer part of the shape. A `CSGCombiner` groups components, so if it's made invisible the expectation is that it's equivalent to making all of its components invisible.

To demonstrate this, I've slightly modified the MRP, where instead of making the `CSGCombiner` invisible its `CSGBox` component is made invisible. The behaviour is the same in 3.2.3.stable and 3.2.4.beta1 and now, with 3.2.4.beta1, the same as making the `CSGCombiner` (or the only shape of a `CSGCombiner`) invisible.
[43251.zip](https://github.com/godotengine/godot/files/5479672/43251.zip)

![43251](https://user-images.githubusercontent.com/9253928/97957963-235be880-1da4-11eb-920d-046f5fe20571.gif)

**Comment by hoontee** [comment]
@madmiraal 

> I think this is the desired behaviour

No, all `CSGShapes` alone will have collision if made invisible. Additionally, collision in `StaticBodies`, `RigidBodies` and `KinematicBodies` will work if invisible.

The reason making the child `CSGBox` invisible in your example makes sense is because the `CSGCombiner` must skip it in its final mesh calculation.

**Comment by hoontee** [comment]
Also, an error when invisible:
![image](https://user-images.githubusercontent.com/5272529/97960879-d9f5a400-1d77-11eb-912b-631a43ad9f7b.png)

**Comment by Janders1800** [comment]
I dont know if that is the desired behaviour but, having invisible CSGShapes collisions is the only way to get around the KinematicBody edge collision bugs; https://github.com/godotengine/godot/issues/34436, if you go for a modular level design approach.

**Comment by madmiraal** [comment]
> No, all `CSGShapes` alone will have collision if made invisible. Additionally, collision in `StaticBodies`, `RigidBodies` and `KinematicBodies` will work if invisible.
> 
> The reason making the child `CSGBox` invisible in your example makes sense is because the `CSGCombiner` must skip it in its final mesh calculation.

It should be consistent. If making a CSG component invisible should only prevent it from been drawn (the [documented](https://docs.godotengine.org/en/stable/classes/class_spatial.html#class-spatial-property-visible) behaviour), but not affect the `CollisionShape`, then the [example](https://github.com/godotengine/godot/files/5479672/43251.zip) I provided [above](https://github.com/godotengine/godot/issues/43251#issuecomment-720948272) shouldn't happen either.

However, currently, visibility does not prevent the CSG component from being drawn. Visibility is used to control which CSG components are included. It affects the mesh directly; resulting in faces being both added and removed depending on the `Operation`. Since a CSG object only has one mesh for both the `VisualInstance` and the `CollisionShape` both are affected.

The way I see it, we have two options:
1. Keep one mesh, collisions only happen with visible CSG components i.e. the bug is that invisible individual CSG Shapes are collidable when they shouldn't be.
2. Have separate meshes for the `VisualInstance` and the `CollisionShape`. The parameters `visible` and `use_collision` could be used to control which CSG components are visible and which are collidable.

**Comment by hoontee** [comment]
> 2. Have separate meshes for the `VisualInstance` and the `CollisionShape`. The parameters `visible` and `use_collision` could be used to control which CSG components are visible and which are collidable.

I would vastly prefer this over the former option considering my current use cases, but it would break compatibility. I will make a new PR for this.

However, for 3.2's sake, are you favorable of merging #40814 or would you prefer the aforementioned PR proposal?

**Comment by hoontee** [comment]
@madmiraal

The PR's ready for submission, just waiting on your response to https://github.com/godotengine/godot/issues/43251#issuecomment-721825015.

![8ed25e99fb2c94ce607ddfbe4bbf87b5](https://user-images.githubusercontent.com/5272529/98175511-60b39980-1ebc-11eb-8741-e926fd2094c8.gif)

**Comment by hoontee** [comment]
Still an issue despite multiple PRs addressing it. Can we decide on whether #40814 or #43381 gets merged?
